File: Exercises_Java/Alarm_System/src/impl/SecurityControlUnit.java
package impl;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.time.LocalTime;

/**
 * Created by lucieburgess on 22/01/2017.
 */
public class SecurityControlUnit extends ControlUnit {

   private List<Sensor> configuredSensors = new ArrayList<>();

    public SecurityControlUnit() {
        super();
    }

    @Override // method has different behaviour to ControlUnit
    public void addSensor(String location, String sensorType) throws IllegalArgumentException, NullPointerException {

        if (location == null || sensorType == null) {
            throw new NullPointerException("You must specify a location and sensor type");
        }
        if (sensorType == "Fire" || sensorType == "Smoke" || sensorType == "") {
            throw new IllegalArgumentException("Sensor type must be a security sensor, such as Motion");

        } else if (sensorType == "Motion") {
            configuredSensors.add(new MotionSensor(location));

        } else {
            System.out.println("You must add a sensor to configure the Control Unit. Please try again");
        }
    }

    public List<Sensor> getSensors() {
        return configuredSensors;
    }

    public boolean timeCheck() {
        LocalTime timePoint = LocalTime.now();
        LocalTime timeOn = LocalTime.of(22,0);
        LocalTime timeOff = LocalTime.of(6,0);
        return (timePoint.isAfter(timeOn) || timePoint.isBefore(timeOff));
    }

    @Override
    public void pollSensors() {
        if (timeCheck()) {
            super.pollSensors();
        }
    }

}

File: exercises/week_02/src/main/scala/SecurityControlUnit.scala
import java.time.LocalTime

import traits.{SecuritySensor, Sensor}

import scala.collection.mutable.ListBuffer

/**
  * Created by lucieburgess on 14/05/2017.
  */
class SecurityControlUnit extends ControlUnit {

  /**
    * Takes a sensor to be configured as an argument
    * @param sensor
    * @return list of configured sensors for this control unit
    */
  def configureSensor(sensor: SecuritySensor): ListBuffer[Sensor] = {
    sensors += sensor
  }

  override def pollSensors() {
    if (timeCheck) super.pollSensors()
  }

  def timeCheck: Boolean = {
    val timePoint = LocalTime.now
    val timeOn = LocalTime.of(22, 0)
    val timeOff = LocalTime.of(6, 0)
    timePoint.isAfter(timeOn) || timePoint.isBefore(timeOff)
  }

}
